src/source-af-xdp.c

```diff
@@ -0,0 +1,944 @@
+
+    ptv->slot = s->slot_next;
+
+    /* Reserve memory for umem  */
+    if (AcquireBuffer(ptv) != TM_ECODE_OK) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    if (AFXDPSocketCreation(ptv) != TM_ECODE_OK) {
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+
+    AFXDPAllThreadsRunning(ptv);
+
+    PacketPoolWait();
+    while (1) {
+        /* Start by checking the state of our interface */
+        if (unlikely(ptv->afxdp_state == AFXDP_STATE_DOWN)) {
+            do {
+                usleep(RECONNECT_TIMEOUT);
+                if (unlikely(suricata_ctl_flags != 0)) {
+                    break;
+                }
+                r = AFXDPTryReopen(ptv);
+            } while (r != TM_ECODE_OK);
+        }
+
+        if (unlikely(suricata_ctl_flags != 0)) {
+            SCLogDebug("Stopping Suricata!");
+            AFXDPDumpCounters(ptv);
+            break;
+        }
+
+        /* Busy polling is not set, using poll() to maintain (relatively) decent
+         * performance. xdp_busy_poll must be disabled for kernels < 5.11
+         */
+        if (!ptv->xsk.enable_busy_poll) {
+            StatsIncr(ptv->tv, ptv->capture_afxdp_poll);
+
+            r = poll(&ptv->xsk.fd, 1, POLL_TIMEOUT);
+
+            /* Report poll results */
+            if (r <= 0) {
+                if (r == 0) {
+                    StatsIncr(ptv->tv, ptv->capture_afxdp_poll_timeout);
+                } else if (r < 0) {
+                    StatsIncr(ptv->tv, ptv->capture_afxdp_poll_failed);
+                    SCLogWarning(SC_ERR_AFXDP_READ, "poll failed with retval %d", r);
+                    AFXDPSwitchState(ptv, AFXDP_STATE_DOWN);
+                }
+
+                DumpStatsEverySecond(ptv, &last_dump);
+                continue;
+            }
+        }
+
+        rcvd = xsk_ring_cons__peek(&ptv->xsk.rx, ptv->xsk.busy_poll_budget, &idx_rx);
+        if (!rcvd) {
+            StatsIncr(ptv->tv, ptv->capture_afxdp_empty_reads);
+            ssize_t ret = WakeupSocket(ptv);
+            if (ret < 0) {
+                SCLogWarning(SC_ERR_AFXDP_READ, "recv failed with retval %ld", ret);
+                AFXDPSwitchState(ptv, AFXDP_STATE_DOWN);
+            }
+            DumpStatsEverySecond(ptv, &last_dump);
+            continue;
+        }
+
+        uint32_t res = xsk_ring_prod__reserve(&ptv->umem.fq, rcvd, &idx_fq);
+        while (res != rcvd) {
+            StatsIncr(ptv->tv, ptv->capture_afxdp_failed_reads);
+            ssize_t ret = WakeupSocket(ptv);
+            if (ret < 0) {
+                SCLogWarning(SC_ERR_AFXDP_READ, "recv failed with retval %ld", ret);
+                AFXDPSwitchState(ptv, AFXDP_STATE_DOWN);
+                continue;
+            }
+            res = xsk_ring_prod__reserve(&ptv->umem.fq, rcvd, &idx_fq);
+        }
+
+        gettimeofday(&ts, NULL);
+        ptv->pkts += rcvd;
+        for (uint32_t i = 0; i < rcvd; i++) {
+            p = PacketGetFromQueueOrAlloc();
+            if (unlikely(p == NULL)) {
```

Comment:
there should probably be a way for the user to observe this edge case. Probably a counter